### PR TITLE
Supervised Bin Normalizer

### PR DESCRIPTION
**Pull request by abgoswam**
Towards #819 

- enabled  `NormalizerMode.SupervisedBinning` 
- added tests

### CODE REVIEW

File: test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs [hunk 1]
```diff
@@ -240,6 +244,87 @@ public void SimpleConstructorsAndExtensions()
...
+            //var trans1 = est1.Fit(data);
+            //var data1 = trans1.Transform(data);
+
+            var est2 = new NormalizingEstimator(Env, NormalizingEstimator.NormalizerMode.MinMax, ("float4", "float4"));
+            var trans2 = est2.Fit(data);
+            var data2 = trans2.Transform(data);
+
+            //var est3 = new NormalizingEstimator(Env, new NormalizingEstimator.MinMaxColumn("float4"));
+            //var trans3 = est3.Fit(data);
+            //var data3 = trans3.Transform(data);
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]: delete

File: test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs [hunk 2]
```diff
@@ -240,6 +244,87 @@ public void SimpleConstructorsAndExtensions()
...
+                    new TextLoader.Column("Label", DataKind.R4, 0),
+                    new TextLoader.Column("float4", DataKind.R4, new[]{new TextLoader.Range(1, 4) }),
+                }
+            });
+
+            var data = loader.Read(dataPath);
+
+            //var est1 = new NormalizingEstimator(Env, "float4");
+            //var trans1 = est1.Fit(data);
+            //var data1 = trans1.Transform(data);
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]: delete

File: test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs [hunk 3]
```diff
@@ -240,6 +244,87 @@ public void SimpleConstructorsAndExtensions()
...
+            data4.Schema.TryGetColumnIndex("float4_1", out int output);
+            using (var cursor = data4.GetRowCursor(col => col == output))
+            {
+                var buffer = default(VBuffer<float>);
+                var getter = cursor.GetGetter<VBuffer<float>>(output);
+                var numRows = 0;
+                while (cursor.MoveNext())
+                {
+                    getter(ref buffer);
+                    numRows += 1;
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]: >numRows [](start = 20, length = 7)

do you need to validate the data?

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs [hunk 4]
```diff
@@ -161,6 +168,29 @@ internal override IColumnFunctionBuilder MakeBuilder(IHost host, int srcIndex, C
...
+        {
+            public readonly int NumBins;
+            public readonly string LabelColumn;
+            public readonly int MinBinSize;
+
+            public SupervisedBinningColumn(string input, string output = null,
+                long maxTrainingExamples = Defaults.MaxTrainingExamples,
+                bool fixZero = true,
+                int numBins = Defaults.NumBins,
+                string labelColumn = "Label",
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Transforms/Normalizer.cs]: >"Label" [](start = 37, length = 7)

DefaultColumnNames.Label

**Comment by Ivanidzo4ka** [comment]
@abgoswam I've looked on code, and it looks decent (despite all comments in tests)
Is tests only reason why it's WIP, or you working on something else?

**Comment by abgoswam** [comment]
@Ivanidzo4ka.  yes, the WIP is because I need to cleanup the test comments.

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs [hunk 5]
```diff
@@ -238,7 +268,7 @@ public SchemaShape GetOutputSchema(SchemaShape inputSchema)
         }
     }
 
-    public sealed partial class NormalizingTransformer : OneToOneTransformerBase
+    public sealed partial class NormalizingTransformer : RowToRowTransformerBase
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Transforms/Normalizer.cs]: >RowToRowTransformerBase [](start = 57, length = 23)

If only place where you need label column is training, and for applying transform you don't need it, you can just leave OneToOneTransfomerBase here.

Sorry if I create confusion, I didn't thought it through properly.

Comment (abgoswam) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: exactly. I was able to get it to work by just making Label column active *during training*, so I was wondering  why we need RowToRowTransformerBase at all.

Will keep it as OneToOneTransformerBase

---
In reply to: [241868788](https://github.com/dotnet/machinelearning/pull/1876#discussion_r241868788) [](ancestors = 241868788)

File: test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs (hunk 3 above)
Comment (abgoswam) [reply test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]: For now I have appended to the existing test `SimpleConstructorsAndExtensions`  to verify schema and values. 

Once this PR goes in, I intent to enable the skipped tests in TLC and possibly port over some TLC tests as well. But that would be a separate PR.

---
In reply to: [241653196](https://github.com/dotnet/machinelearning/pull/1876#discussion_r241653196) [](ancestors = 241653196)

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs [hunk 6]
```diff
@@ -161,6 +168,29 @@ internal override IColumnFunctionBuilder MakeBuilder(IHost host, int srcIndex, C
                 => NormalizeTransform.BinUtils.CreateBuilder(this, host, srcIndex, srcType, cursor);
         }
 
+        public sealed class SupervisedBinningColumn : FixZeroColumnBase
```
[the remark this answers is not in the document]
Comment (abgoswam) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: sure. i will look into adding a sample for this  (separate PR)

---
In reply to: [242284487](https://github.com/dotnet/machinelearning/pull/1876#discussion_r242284487) [](ancestors = 242284487)

**Pull request merged**